client/util.go

```diff
@@ -12,55 +12,57 @@ import (
 
 // pollingWatcher generalizes the `Watch` interface for clients which learn new values
 // by asking for them once each group period.
-func pollingWatcher(ctx context.Context, client Client, group *key.Group, log log.Logger) <-chan Result {
-	ch := make(chan Result, 1)
-	r := client.RoundAt(time.Now())
-	val, err := client.Get(ctx, r)
-	if err != nil {
-		log.Error("polling_client", "failed to watch", "err", err)
-		close(ch)
-		return ch
-	}
-	ch <- val
-
-	go func() {
-		defer close(ch)
-
-		// Initially, wait to synchronize to the round boundary.
-		_, nextTime := beacon.NextRound(time.Now().Unix(), group.Period, group.GenesisTime)
-		select {
-		case <-ctx.Done():
-			return
-		case <-time.After(time.Duration(nextTime-time.Now().Unix()) * time.Second):
-		}
-
-		r, err := client.Get(ctx, client.RoundAt(time.Now()))
-		if err == nil {
-			ch <- r
-		} else {
+func pollingWatcher(log log.Logger) GetNotifierFunc {
+	return func(ctx context.Context, client Client, group *key.Group) <-chan Result {
+		ch := make(chan Result, 1)
+		r := client.RoundAt(time.Now())
+		val, err := client.Get(ctx, r)
+		if err != nil {
 			log.Error("polling_client", "failed to watch", "err", err)
+			close(ch)
+			return ch
 		}
+		ch <- val
 
-		// Then tick each period.
-		t := time.NewTicker(group.Period)
-		defer t.Stop()
-		for {
+		go func() {
+			defer close(ch)
+
+			// Initially, wait to synchronize to the round boundary.
+			_, nextTime := beacon.NextRound(time.Now().Unix(), group.Period, group.GenesisTime)
 			select {
-			case <-t.C:
-				r, err := client.Get(ctx, client.RoundAt(time.Now()))
-				if err == nil {
-					ch <- r
-				} else {
-					log.Error("polling_client", "failed to watch", "err", err)
-				}
-				// TODO: keep trying on errors?
 			case <-ctx.Done():
 				return
+			case <-time.After(time.Duration(nextTime-time.Now().Unix()) * time.Second):
 			}
-		}
-	}()
 
-	return ch
+			r, err := client.Get(ctx, client.RoundAt(time.Now()))
+			if err == nil {
+				ch <- r
+			} else {
+				log.Error("polling_client", "failed to watch", "err", err)
+			}
+
+			// Then tick each period.
+			t := time.NewTicker(group.Period)
+			defer t.Stop()
+			for {
+				select {
+				case <-t.C:
+					r, err := client.Get(ctx, client.RoundAt(time.Now()))
```

Comment:
If for whatever reason client.Get blocks for more than "group.Period" seconds, the next time we enter the for loop we will be asking for the current round (because "client.RoundAt(time.Now())") which will cause us to miss a round in between.

I think this is ok, but we should call it out in the PR that if "network conditions slow down the delivery of a round, the client will skip as many rounds as the duration of the network condition".

Reply:
Agree, I'll add it to the description and make sure it's copied into the commit description when this is squashed and merged.

FWIW I did consider making this more complicated to cater for this case but I didn't get far before I decided to back out and just go with the simple case for this PR.